
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Welcome to PolyA’s documentation! &#8212; PolyA 0.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="polyA" href="modules.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="welcome-to-polya-s-documentation">
<h1>Welcome to PolyA’s documentation!<a class="headerlink" href="#welcome-to-polya-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules.html">polyA</a></li>
</ul>
</div>
</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
<section id="readme">
<h1>Readme<a class="headerlink" href="#readme" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://github.com/TravisWheelerLab/polyA/workflows/PolyA%20CI/badge.svg"><img alt="PolyA CI" src="https://github.com/TravisWheelerLab/polyA/workflows/PolyA%20CI/badge.svg" /></a>
<a class="reference external image-reference" href="https://img.shields.io/github/license/TravisWheelerLab/polyA"><img alt="GitHub" src="https://img.shields.io/github/license/TravisWheelerLab/polyA" /></a>
<a class="reference external image-reference" href="https://img.shields.io/pypi/v/polyA"><img alt="PyPI" src="https://img.shields.io/pypi/v/polyA" /></a>
</section>
<section id="aaaaaaaaaaaaaaaa-polya">
<h1>AAAAAAAAAAAAAAAA (PolyA):<a class="headerlink" href="#aaaaaaaaaaaaaaaa-polya" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p><strong>A</strong>utomatically <strong>A</strong>djudicate <strong>A</strong>ny <strong>A</strong>nd <strong>A</strong>ll <strong>A</strong>rbitrary
<strong>A</strong>nnotations, <strong>A</strong>stutely <strong>A</strong>djoin <strong>A</strong>butting <strong>A</strong>lignments,
<strong>A</strong>nd <strong>A</strong>lso <strong>A</strong>mputate <strong>A</strong>nything <strong>A</strong>miss</p>
</div></blockquote>
<p>A tool for adjudicating between competing annotations of biological sequences.</p>
<section id="about">
<h2>About<a class="headerlink" href="#about" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/2021.02.13.430877v1">Preprint on
bioRxiv</a></p>
<p>Annotation of a biological sequence is usually performed by aligning that
sequence to a database of known sequence elements. When that database contains
elements that are highly similar to each other, the proper annotation may
be ambiguous, because several entries in the database produce high-scoring
alignments. Typical annotation methods work by assigning a label based
on the candidate annotation with the highest alignment score; this can
overstate annotation certainty, mislabel boundaries, and fails to identify
large scale rearrangements or insertions within the annotated sequence.</p>
<p>PolyA is a software tool that adjudicates between competing alignment-based
annotations by computing estimates of annotation confidence, identifying a
trace with maximal confidence, and recursively splicing/stitching inserted
elements. PolyA communicates annotation certainty, identifies large scale
rearrangements, and detects boundaries between neighboring elements.</p>
</section>
<section id="using">
<h2>Using<a class="headerlink" href="#using" title="Permalink to this headline">¶</a></h2>
<p>PolyA may be consumed as an ordinary Python package:</p>
<p>Install using the <code class="docutils literal notranslate"><span class="pre">pip</span></code> tool:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">polyA</span>
</pre></div>
</div>
<p>Run from the command line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">polyA</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>or</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">polyA</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<section id="command-line">
<h3>Command Line<a class="headerlink" href="#command-line" title="Permalink to this headline">¶</a></h3>
<p>Command line usage is available with <code class="docutils literal notranslate"><span class="pre">polyA</span> <span class="pre">-h</span></code>. It is also included below for
convenience.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">polyA</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">chunk</span><span class="o">-</span><span class="n">size</span> <span class="n">CHUNK_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">confidence</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">prior</span><span class="o">-</span><span class="n">counts</span> <span class="n">FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">shard</span><span class="o">-</span><span class="n">gap</span> <span class="n">SHARD_GAP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sequences</span> <span class="n">SEQS</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">ultra</span><span class="o">-</span><span class="n">data</span> <span class="n">FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">easel</span><span class="o">-</span><span class="n">path</span> <span class="n">BIN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ultra</span><span class="o">-</span><span class="n">path</span> <span class="n">BIN</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">file</span> <span class="n">LOG</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">level</span> <span class="n">LEVEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">matrix</span><span class="o">-</span><span class="n">position</span><span class="p">]</span>
             <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">path</span> <span class="n">PATH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sequence</span><span class="o">-</span><span class="n">position</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">soda</span><span class="p">]</span>
             <span class="n">ALIGNMENTS</span> <span class="n">MATRICES</span>

<span class="n">PolyA</span> <span class="n">sequence</span> <span class="n">adjudication</span> <span class="n">tool</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">ALIGNMENTS</span>              <span class="n">alignments</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">Stockholm</span> <span class="nb">format</span>
  <span class="n">MATRICES</span>                <span class="n">substitution</span> <span class="n">matrices</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">PolyA</span> <span class="n">matrix</span> <span class="nb">format</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>              <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">-</span><span class="n">v</span><span class="p">,</span> <span class="o">--</span><span class="n">version</span>           <span class="n">show</span> <span class="n">version</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">chunk</span><span class="o">-</span><span class="n">size</span> <span class="n">CHUNK_SIZE</span>
                          <span class="n">size</span> <span class="n">of</span> <span class="n">the</span> <span class="n">window</span> <span class="ow">in</span> <span class="n">base</span> <span class="n">pairs</span> <span class="n">analyzed</span> <span class="n">together</span>
  <span class="o">--</span><span class="n">confidence</span>            <span class="n">run</span> <span class="n">the</span> <span class="n">confidence</span> <span class="n">calculation</span> <span class="ow">and</span> <span class="n">then</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">prior</span><span class="o">-</span><span class="n">counts</span> <span class="n">FILE</span>     <span class="n">file</span> <span class="n">containing</span> <span class="n">query</span> <span class="n">genomic</span> <span class="n">counts</span>
  <span class="o">--</span><span class="n">shard</span><span class="o">-</span><span class="n">gap</span> <span class="n">SHARD_GAP</span>
                          <span class="n">maximum</span> <span class="n">alignment</span> <span class="n">gap</span> <span class="n">before</span> <span class="n">sharding</span> <span class="n">occurs</span>
  <span class="o">--</span><span class="n">sequences</span> <span class="n">SEQS</span>        <span class="n">FASTA</span> <span class="n">file</span> <span class="n">of</span> <span class="n">the</span> <span class="n">target</span> <span class="n">sequence</span> <span class="k">for</span> <span class="n">using</span> <span class="n">ULTRA</span>
  <span class="o">--</span><span class="n">ultra</span><span class="o">-</span><span class="n">data</span> <span class="n">FILE</span>       <span class="n">text</span> <span class="n">file</span> <span class="n">of</span> <span class="n">the</span> <span class="n">output</span> <span class="kn">from</span> <span class="nn">ULTRA</span> <span class="n">ran</span> <span class="n">on</span> <span class="n">the</span> <span class="n">FASTA</span> <span class="n">file</span>
                          <span class="n">of</span> <span class="n">the</span> <span class="n">target</span> <span class="n">sequence</span>
  <span class="o">--</span><span class="n">easel</span><span class="o">-</span><span class="n">path</span> <span class="n">BIN</span>        <span class="n">path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">esl_scorematrix</span> <span class="n">program</span><span class="p">,</span> <span class="k">if</span> <span class="n">necessary</span>
                          <span class="p">(</span><span class="n">assumed</span> <span class="n">to</span> <span class="n">be</span> <span class="ow">in</span> <span class="n">PATH</span><span class="p">)</span>
  <span class="o">--</span><span class="n">ultra</span><span class="o">-</span><span class="n">path</span> <span class="n">BIN</span>        <span class="n">path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">ULTRA</span> <span class="n">binary</span> <span class="n">to</span> <span class="n">use</span><span class="p">,</span> <span class="k">if</span> <span class="n">necessary</span> <span class="p">(</span><span class="n">assumed</span>
                          <span class="n">to</span> <span class="n">be</span> <span class="ow">in</span> <span class="n">PATH</span><span class="p">)</span>
  <span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">file</span> <span class="n">LOG</span>          <span class="n">file</span> <span class="n">to</span> <span class="n">store</span> <span class="n">log</span> <span class="n">output</span> <span class="ow">in</span><span class="p">,</span> <span class="n">defaults</span> <span class="n">to</span> <span class="n">stderr</span>
  <span class="o">--</span><span class="n">log</span><span class="o">-</span><span class="n">level</span> <span class="n">LEVEL</span>       <span class="n">logging</span> <span class="n">level</span> <span class="n">to</span> <span class="n">use</span><span class="p">,</span> <span class="s1">&#39;debug&#39;</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">most</span> <span class="n">noisy</span>
  <span class="o">--</span><span class="n">matrix</span><span class="o">-</span><span class="n">position</span>       <span class="n">produce</span> <span class="n">output</span> <span class="ow">in</span> <span class="n">terms</span> <span class="n">of</span> <span class="n">the</span> <span class="n">matrix</span> <span class="n">position</span>
  <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">path</span> <span class="n">PATH</span>      <span class="n">directory</span> <span class="n">to</span> <span class="n">write</span> <span class="n">output</span> <span class="n">files</span> <span class="n">to</span><span class="p">,</span> <span class="n">defaults</span> <span class="n">to</span>
                          <span class="n">working</span> <span class="n">directory</span>
  <span class="o">--</span><span class="n">sequence</span><span class="o">-</span><span class="n">position</span>     <span class="n">produce</span> <span class="n">output</span> <span class="ow">in</span> <span class="n">terms</span> <span class="n">of</span> <span class="n">the</span> <span class="n">target</span> <span class="n">sequence</span>
                          <span class="n">position</span>
  <span class="o">--</span><span class="n">soda</span>                  <span class="n">write</span> <span class="n">a</span> <span class="n">SODA</span> <span class="n">visualization</span> <span class="n">file</span> <span class="n">to</span> <span class="n">the</span> <span class="n">output</span>
                          <span class="n">directory</span>
  <span class="o">--</span><span class="n">complexity</span><span class="o">-</span><span class="n">adjustment</span> <span class="n">complexity</span><span class="o">-</span><span class="n">adjust</span> <span class="n">alignment</span> <span class="n">scores</span> <span class="p">(</span><span class="n">scores</span> <span class="n">of</span> <span class="n">matches</span>
                          <span class="n">between</span> <span class="n">sequence</span> <span class="n">regions</span> <span class="n">of</span> <span class="n">biassed</span> <span class="n">nucleotide</span>
                          <span class="n">composition</span> <span class="n">are</span> <span class="n">adjusted</span> <span class="n">downwards</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="input-formats">
<h3>Input Formats<a class="headerlink" href="#input-formats" title="Permalink to this headline">¶</a></h3>
<p>PolyA accepts two required inputs and several optional inputs that affect
its behavior. The required inputs are an alignment file which must contain
alignments for all possible queries matching the target sequence. This file
must be in <a class="reference external" href="https://sonnhammer.sbc.su.se/Stockholm.html">Stockholm</a> format
with several custom metadata fields. The other required input is a set of
substitution matrices. This file uses a custom, but extremely simple format.</p>
<section id="alignment-file-format">
<h4>Alignment File Format<a class="headerlink" href="#alignment-file-format" title="Permalink to this headline">¶</a></h4>
<p>Alignments for all possible queries matching the target sequence should be
contained in a single file in Stockholm format.</p>
<p>There are several special metadata fields that must exist for each alignment in
this file. See the example below. An explanation is indented to the right of
each field with additional detail as noted.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#=GF ID  MERX#DNA/TcMar-Tigger    query sequence (1)</span>
<span class="c1">#=GF TR  chr1:11543-28567         target sequence</span>
<span class="c1">#=GF SC  1153                     alignment score</span>
<span class="c1">#=GF SD  +                        strand</span>
<span class="c1">#=GF TQ  -1                       (2)</span>
<span class="c1">#=GF ST  127                      alignment start position on target</span>
<span class="c1">#=GF SP  601                      alignment stop position on target</span>
<span class="c1">#=GF CST 135                      alignment start position on query</span>
<span class="c1">#=GF CSP 628                      alignment stop position on query</span>
<span class="c1">#=GF FL  128                      (3)</span>
<span class="c1">#=GF MX  matrix_name              (4)</span>
<span class="c1">#=GF GI -25                       gap init</span>
<span class="c1">#=GF GE -5                        gap extension</span>
</pre></div>
</div>
<ul class="simple">
<li><ol class="arabic simple">
<li><p>query sequence names must be in the format ‘name#family/class’</p></li>
</ol>
</li>
<li><p>(2) valid values: ‘q’ if the alignment is on the reverse
strand and the reversed sequence is the query; ‘t’ if the alignment
is on the reverse strand and the reversed sequence is the target;
‘-1’ if the alignment is on the positive strand</p></li>
<li><ol class="arabic simple" start="3">
<li><p>the flanking region of the unaligned query sequence</p></li>
</ol>
</li>
<li><ol class="arabic simple" start="4">
<li><p>the name of the substitution matrix file used to create alignment</p></li>
</ol>
</li>
</ul>
</section>
<section id="creating-alignment-files-from-cross-match-alignments">
<h4>Creating Alignment files from cross_match alignments<a class="headerlink" href="#creating-alignment-files-from-cross-match-alignments" title="Permalink to this headline">¶</a></h4>
<p>We include a parser to convert cross_match alignment files to stockholm
alignments. This can be executed with the <code class="docutils literal notranslate"><span class="pre">cm_to_stockholm</span></code> script (installed
with PolyA) or through the <code class="docutils literal notranslate"><span class="pre">polyA.converters</span></code> module (<code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">-m</span>
<span class="pre">polyA.converters.cm_to_stockholm</span></code>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">cm_to_stockholm</span> <span class="n">align_file</span><span class="o">.</span><span class="n">cm</span>
<span class="n">outputs</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="nb">input</span> <span class="n">directly</span> <span class="n">into</span> <span class="n">polyA</span><span class="p">):</span>
    <span class="n">align_file</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">sto</span>
    <span class="n">align_file</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">matrix</span>
</pre></div>
</div>
</section>
<section id="substitution-matrix-files">
<h4>Substitution Matrix Files<a class="headerlink" href="#substitution-matrix-files" title="Permalink to this headline">¶</a></h4>
<p>Substitution matrix file example format (can include ambiguity codes):</p>
<ul class="simple">
<li><p>this file must include all of the matrices specified in the “#=GF MX” field of the alignment file, with corresponding and matching matrix names</p></li>
<li><p>if lambda is not included polyA will use esl_scorematrix to calculate it for all matrices</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">matrix_name</span> <span class="k">lambda</span><span class="p">(</span><span class="n">optional</span><span class="p">)</span>
  <span class="n">A</span>   <span class="n">G</span>   <span class="n">C</span>    <span class="n">T</span>    <span class="n">N</span>
  <span class="mi">8</span>  <span class="o">-</span><span class="mi">6</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">15</span>  <span class="o">-</span><span class="mi">1</span>
 <span class="o">-</span><span class="mi">2</span>  <span class="mi">10</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">1</span>
<span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">13</span>  <span class="mi">10</span>  <span class="o">-</span><span class="mi">2</span>   <span class="o">-</span><span class="mi">1</span>
<span class="o">-</span><span class="mi">15</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">6</span>   <span class="mi">8</span>   <span class="o">-</span><span class="mi">1</span>
 <span class="o">-</span><span class="mi">1</span>  <span class="o">-</span><span class="mi">1</span>   <span class="o">-</span><span class="mi">1</span>  <span class="o">-</span><span class="mi">1</span>   <span class="o">-</span><span class="mi">1</span>
<span class="o">//</span>
<span class="n">matrix_name2</span> <span class="n">lambda2</span><span class="p">(</span><span class="n">optional</span><span class="p">)</span>
  <span class="n">A</span>   <span class="n">G</span>   <span class="n">C</span>    <span class="n">T</span>    <span class="n">N</span>
  <span class="mi">8</span>  <span class="o">-</span><span class="mi">6</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">15</span>  <span class="o">-</span><span class="mi">1</span>
 <span class="o">-</span><span class="mi">2</span>  <span class="mi">10</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">1</span>
<span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">13</span>  <span class="mi">10</span>  <span class="o">-</span><span class="mi">2</span>   <span class="o">-</span><span class="mi">1</span>
<span class="o">-</span><span class="mi">15</span>  <span class="o">-</span><span class="mi">13</span>  <span class="o">-</span><span class="mi">6</span>   <span class="mi">8</span>   <span class="o">-</span><span class="mi">1</span>
 <span class="o">-</span><span class="mi">1</span>  <span class="o">-</span><span class="mi">1</span>   <span class="o">-</span><span class="mi">1</span>  <span class="o">-</span><span class="mi">1</span>   <span class="o">-</span><span class="mi">1</span>
<span class="o">//</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
<section id="sequence-file-optional">
<h4>Sequence File (Optional)<a class="headerlink" href="#sequence-file-optional" title="Permalink to this headline">¶</a></h4>
<p>A FASTA file of the target sequence is needed when using ULTRA.
The target sequence must be the same genomic region that was used
to get the cross_match alignment file. This file must follow the
format of</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">chrom</span><span class="p">:</span><span class="n">start</span><span class="o">-</span><span class="n">end</span>
<span class="n">target_sequence</span>
</pre></div>
</div>
<p>as shown in the example.</p>
<p>Sequence file example format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">chr1</span><span class="p">:</span><span class="mi">152302175</span><span class="o">-</span><span class="mi">152325203</span>
<span class="n">AATAGTTTATTTTTAATTTAGATGCAGCTTACTATAATATTAATTATGTCCAAGATGATT</span>
<span class="n">TTTTGAATACAGAATACTAGAATTCCAATAGAAGGATAATAGAGAAAGATGTGCTAGCCC</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
</section>
<section id="output-formats">
<h3>Output Formats<a class="headerlink" href="#output-formats" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">start</span>   <span class="n">stop</span>    <span class="n">ID</span>  <span class="n">name</span>
<span class="o">----------------------------------------</span>
<span class="mi">11990879</span>    <span class="mi">11991268</span>    <span class="n">eaa042dd09f944f68dba2fd4727c64e2</span>    <span class="n">LTR40a</span><span class="c1">#LTR/ERVL</span>
<span class="mi">11991272</span>    <span class="mi">11991444</span>    <span class="n">fb5ef5e0e2ca4e05837ddc34ca7ef9e4</span>    <span class="n">MSTA1</span><span class="c1">#LTR/ERVL-MaLR</span>
<span class="mi">11991445</span>    <span class="mi">11991562</span>    <span class="n">bdfc4039b7d947d0b25bf1115cc282ed</span>    <span class="n">AluJr4</span><span class="c1">#SINE/Alu</span>
<span class="mi">11991563</span>    <span class="mi">11991573</span>    <span class="mi">4871</span><span class="n">d91441a146209b98f645feae68c8</span>    <span class="n">FLAM_C</span><span class="c1">#SINE/Alu</span>
<span class="mi">11991574</span>    <span class="mi">11991818</span>    <span class="n">fb5ef5e0e2ca4e05837ddc34ca7ef9e4</span>    <span class="n">MSTB1</span><span class="c1">#LTR/ERVL-MaLR</span>
<span class="mi">11991819</span>    <span class="mi">11991875</span>    <span class="n">eaa042dd09f944f68dba2fd4727c64e2</span>    <span class="n">LTR40a</span><span class="c1">#LTR/ERVL</span>

<span class="o">*</span> <span class="n">Matching</span> <span class="n">IDnums</span> <span class="n">correspond</span> <span class="n">to</span> <span class="n">partial</span> <span class="n">sequences</span> <span class="n">that</span> <span class="n">originate</span> <span class="kn">from</span>
<span class="nn">the</span> <span class="n">same</span> <span class="n">ancestral</span> <span class="n">sequence</span><span class="o">.</span>
</pre></div>
</div>
<section id="confidence-only-output-file-format">
<h4>Confidence only output file format<a class="headerlink" href="#confidence-only-output-file-format" title="Permalink to this headline">¶</a></h4>
<p>Computes confidence of a single input alignment region. Does not perform
annotation or adjudication, simply outputs the confidence of all competing
queries given in the input.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">query_label</span>         <span class="n">confidence</span>
<span class="n">LTR40a</span><span class="c1">#LTR/ERVL     0.875</span>
<span class="n">LTR40b</span><span class="c1">#LTR/ERVL     0.052</span>
<span class="n">LTR40c</span><span class="c1">#LTR/ERVL     0.001</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
</section>
<section id="extensions">
<h3>Extensions<a class="headerlink" href="#extensions" title="Permalink to this headline">¶</a></h3>
<section id="visualizing-annotations-using-soda">
<h4>Visualizing annotations using SODA<a class="headerlink" href="#visualizing-annotations-using-soda" title="Permalink to this headline">¶</a></h4>
<p><span class="raw-html-m2r"><!--
TODO(Audrey): Add updated user explanation
The command line option --soda will output the annotation data to a single json file
to be used in SODA.
The visualization will also give information about the original alignment that each output
family annotation belongs to.
--></span>
This section is a work in progress and will be released in the coming weeks.</p>
</section>
<section id="prior-counts-files">
<h4>Prior Counts Files<a class="headerlink" href="#prior-counts-files" title="Permalink to this headline">¶</a></h4>
<p>Default confidence calculations assume a uniform distribution over all
competing queries. In the case of non uniform priors, the command line option –prior-counts prior_counts.txt includes prior
genome counts in confidence calculations (see paper for more details).</p>
<p>TODO(Kaitlin): Add link to paper</p>
<p>Prior counts file example format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">subfamily</span>   <span class="n">genome_count</span>
<span class="n">AluYk2</span>      <span class="mi">6855</span>
<span class="n">LTR38</span>       <span class="mi">255</span>
<span class="n">L1PA7_5end</span>  <span class="mi">13261</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
<section id="using-ultra">
<h4>Using ULTRA<a class="headerlink" href="#using-ultra" title="Permalink to this headline">¶</a></h4>
<p>The optional use of ULTRA allows polyA to include tandem repeats (TRs) in the competing annotations
of the target sequence. Doing so removes the dependency on pre-masking TRs prior to annotation, allows
TRs to outcompete potentially weak fragmentary family annotation , and allows a family annotation
to outcompete a TR.
The command line option –sequences seq.fasta (with –ultra-path if necessary) will
run ULTRA with polyA or –ultra-data ultra_data.txt can be used if ULTRA was ran on seq.fasta prior.</p>
</section>
</section>
<section id="additional-software">
<h3>Additional software<a class="headerlink" href="#additional-software" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">esl_scorematrix</span></code> utility is a part of the Easel package in the <a class="reference external" href="http://hmmer.org">HMMER
software suite</a>. Its source code is available at
[<a class="reference external" href="https://github.com/EddyRivasLab/easel/">https://github.com/EddyRivasLab/easel/</a>].</p>
<p>We use it to compute a <code class="docutils literal notranslate"><span class="pre">lambda</span></code> value for each score matrix, unless one is given
in the matrix file. The easiest way to run the utility, either manually or as
part of the PolyA pipeline is to use the <a class="reference external" href="https://hub.docker.com/r/traviswheelerlab/polya-esl_scorematrix">Docker
container</a> we
provide.</p>
<p>It is then possible to run <code class="docutils literal notranslate"><span class="pre">esl_scorematrix</span></code> with the following command (which
just shows the help information):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">mount</span> <span class="n">src</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">target</span><span class="o">=/</span><span class="n">data</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="n">bind</span> <span class="n">traviswheelerlab</span><span class="o">/</span><span class="n">polya</span><span class="o">-</span><span class="n">esl_scorematrix</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>Paths passed to the command will need to be relative to <code class="docutils literal notranslate"><span class="pre">/data</span></code>, which is
mounted as the current working directory outside the container.</p>
<p><a class="reference external" href="https://github.com/TravisWheelerLab/ultra">ULTRA</a></p>
<ul class="simple">
<li><p>will detect tandem repeat regions in a sequence FASTA file</p></li>
</ul>
</section>
</section>
<section id="development">
<h2>Development<a class="headerlink" href="#development" title="Permalink to this headline">¶</a></h2>
<p>TODO(George): Mention flit install and the –symlink option (add a make target)</p>
<p>This project uses <a class="reference external" href="https://pipenv.pypa.io/en/latest/">Pipenv</a>, which can be
installed through Homebrew for Mac users. It must be installed before the
Makefile targets, or the other commands listed in this document will work.</p>
<p>In order to run a command which relies on the project virtual environment, such
as <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">foo.py</span></code>, it is necessary to either run <code class="docutils literal notranslate"><span class="pre">pipenv</span> <span class="pre">shell</span></code> first, which
will put you into a shell that has the correct Python in its <code class="docutils literal notranslate"><span class="pre">PATH</span></code>, or prefix
the command with <code class="docutils literal notranslate"><span class="pre">pipenv</span> <span class="pre">run</span></code> (e.g. <code class="docutils literal notranslate"><span class="pre">pipenv</span> <span class="pre">run</span> <span class="pre">python</span> <span class="pre">foo.py</span></code>).</p>
<section id="makefile">
<h3>Makefile<a class="headerlink" href="#makefile" title="Permalink to this headline">¶</a></h3>
<p>A Makefile is available, run <code class="docutils literal notranslate"><span class="pre">make</span></code> or <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">help</span></code> in the project root to see
the available targets.</p>
</section>
<section id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h3>
<p>If you prefer not to use the Makefile, or if you need to add or remove
dependencies, the following commands will allow you to manage dependencies.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get to work (from within project directory)</span>
<span class="c1"># This drops you into a shell inside the project virtual environment which means</span>
<span class="c1"># that commands that &quot;pipenv run&quot; may be elided from other commands.</span>
<span class="n">pipenv</span> <span class="n">shell</span>

<span class="c1"># Fetch runtime dependencies</span>
<span class="n">pipenv</span> <span class="n">install</span>

<span class="c1"># Fetch runtime and development dependencies</span>
<span class="n">pipenv</span> <span class="n">install</span> <span class="o">--</span><span class="n">dev</span>

<span class="c1"># Add a runtime dependency</span>
<span class="n">pipenv</span> <span class="n">install</span> <span class="o">&lt;</span><span class="n">package</span><span class="o">&gt;</span>

<span class="c1"># Add a development dependency</span>
<span class="n">pipenv</span> <span class="n">install</span> <span class="o">--</span><span class="n">dev</span> <span class="o">&lt;</span><span class="n">package</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="docker-image">
<h3>Docker Image<a class="headerlink" href="#docker-image" title="Permalink to this headline">¶</a></h3>
<p>There is a <code class="docutils literal notranslate"><span class="pre">Dockerfile</span></code> in the repo root. The image it describes is
used for running tests in CI and can be used locally for convenience.</p>
<p>When dependencies change the image must be rebuilt and the new
version pushed to Docker Hub. This can be done with <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">container</span></code>
if you have the correct permissions. Otherwise, ask a maintainer
to do it for you.</p>
</section>
<section id="unit-tests">
<h3>Unit Tests<a class="headerlink" href="#unit-tests" title="Permalink to this headline">¶</a></h3>
<p>Unit tests use <a class="reference external" href="https://pytest.org/en/latest/">pytest</a>. The linked
documentation contains examples of how to use it to both write and run tests.
The <a class="reference external" href="https://pytest.org/en/latest/example/index.html">examples</a> are quite
extensive.</p>
<p>Documentation tests are also fine for simple cases.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run tests</span>
<span class="n">pipenv</span> <span class="n">run</span> <span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">pytest</span>

<span class="c1"># or</span>
<span class="n">make</span> <span class="n">check</span><span class="o">-</span><span class="n">fast</span> <span class="n">check</span><span class="o">-</span><span class="n">slow</span>

<span class="c1"># or</span>
<span class="n">make</span> <span class="n">check</span>
</pre></div>
</div>
</section>
<section id="documentation">
<h3>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h3>
<p>We use Sphinx for project documentation. Run <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">docs</span></code> to update the
documentation sources and build the HTML. Use <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">docs-serve</span></code> to serve the
documentation locally.</p>
</section>
</section>
<section id="license">
<h2>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h2>
<p>BSD license. See <code class="docutils literal notranslate"><span class="pre">LICENSE</span></code>.</p>
</section>
<section id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://wheelerlab.org">Wheeler Lab</a> at the University of Montana.</p>
<ul class="simple">
<li><p>Kaitlin Carey</p></li>
<li><p>Audrey Shingleton</p></li>
<li><p>George Lesica</p></li>
<li><p>Jack Roddy</p></li>
<li><p>Travis Wheeler</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">PolyA</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules.html">polyA</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="modules.html" title="next chapter">polyA</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Audrey Shingleton, Kaitlin Carey, George Lesica, Travis Wheeler.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>